File: Competitive Programmers Handbook/Introduction/R.py
import math

memo = {0: 0}
def pre():
    numDigits = 1
    last_pos = 0
    for numDigits in range(1, 21):
        last_pos += 9 * (10**(numDigits-1)) * numDigits
        memo[numDigits] = last_pos

def numDigits(query):
    for digit in range(1, 21):
        if memo[digit] >= query:
            return digit

def solve(query):
    digits = numDigits(query)
    num_start = (10 ** (digits-1))
    num_end = int(10 ** digits) - 1
    range_start = memo[digits-1] + 1
    range_end = memo[digits]

    _num_start = num_start
    _num_end = num_end
    _range_start = range_start
    _range_end = range_end
    mid = 0
    while True:
        mid = int((_num_start + _num_end) / 2)
        _range_mid = _range_start + ((mid - _num_start) * digits)
        if (query >= _range_mid):
            _num_start = mid
            _range_start = _range_mid
        else:
            _num_end = mid
        if (abs(query - _range_start) < 1000):
            break
    range_diff = (query - _range_start)
    diff = int(range_diff / digits)
    num = _num_start + diff
    pos = query - (_range_start + (diff * digits))
    num = str(num)
    return num[pos]


pre()
num_queries = int(input())

for i in range(num_queries):
    query = int(input())
    print(solve(query))

File: Competitive Programmers Handbook/Introduction/R.cpp
#include <bits/stdc++.h>

using namespace std;

#if defined(LOCALCODE) && !defined(ONLINE_JUDGE)
#define TEST
#endif
typedef long long ll;
typedef vector<int> vi;
typedef vector<ll> vll;
#define mp make_pair
#define FOR(i, j, k, inc) for (int i = j; i < k; i += inc)
#define REP(i, n) for (int i = 0; i < n; i++)
#ifdef TEST
#define debug(x...) cerr << #x << x << endl;
#else
#define debug(x...)
#endif

ll pos_memo[18] = {0};
ll max_position(ll num_digits)
{
    ll ans = (num_digits * pow(10, num_digits)) - (pow(10, num_digits) - 10) / 9 - 1;
    return ans;
}

void pre()
{
    FOR(i, 1, 18, 1)
    {
        pos_memo[i] = max_position(i);
    }
}

int num(ll req_pos, ll numDigits)
{
    if (!(req_pos <= pos_memo[numDigits] && req_pos >= pos_memo[numDigits - 1]))
    {
        throw "Baka move somewhere";
    }
    ll range_start = pow(10, numDigits - 1);
    // Round off error on bigger numbers when using pow(10, numDigits) -1
    ll range_end = ((ll)pow(10, numDigits)) - 1;

    ll digitsuptolast = pos_memo[numDigits - 1];
    ll pos_diff = req_pos - digitsuptolast;
    ll num_diff = pos_diff / numDigits;

    // When pos_diff > 1e17 num_diff is wrong
    // if (pos_diff < 1e17)
    // {
    //     if (pos_diff % numDigits == 0)
    //     {
    //         ll num = range_start + num_diff - 1;
    //         return num;
    //     }
    //     else
    //     {
    //         ll num = range_start + num_diff;
    //         return num;
    //     }
    // }

    ll num_start = 0,
       num_end = 9 * (ll)pow(10, numDigits - 1),
       num_mid = (num_start + num_end + 1) / 2;

    ll digits_start = 0,
       digits_end = num_end * numDigits,
       digits_mid = 0;
    req_pos -= digitsuptolast;
    while (true)
    {
        if (num_end - num_start + 1 < 100)
        {
            break;
        }
        num_mid = (num_start + num_end + 1) / 2;
        digits_mid = digits_start + (num_mid - num_start) * numDigits;
        if (digits_mid <= req_pos)
        {
            num_start = num_mid;
            digits_start = digits_mid;
        }
        else
        {
            num_end = num_mid;
            digits_end = digits_mid;
        }
    }
    // digits_start -= numDigits;
    pos_diff = req_pos - digits_start;
    num_diff = (pos_diff / numDigits);
    ll num = range_start + num_start + num_diff;
    stringstream ss;
    ss << num;
    string num_s;
    ss >> num_s;
    int index = (pos_diff % numDigits);
    int ans = (num_s[index] - '0');
    return index;
}

int solve(ll pos)
{
    ll num_digits = 0;
    FOR(i, 1, 18, 1)
    {
        if (pos_memo[i] >= pos)
        {
            num_digits = i;
            break;
        }
    }

    int number = num(pos, num_digits);
    return number;
}

int main()
{
#ifdef TEST
    freopen("/media/psuv/Data/CP/input.in", "r", stdin);
    freopen("/media/psuv/Data/CP/output.out", "w", stdout);
    freopen("/media/psuv/Data/CP/output.out", "w", stderr);
    auto start = chrono::high_resolution_clock::now();
#endif
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    pre();

    int q;
    cin >> q;
    REP(i, q)
    {
        ll pos;
        cin >> pos;
        auto ans = solve(pos);
        cout << ans << "\n";
    }

#ifdef TEST
    auto end = chrono::high_resolution_clock::now();
    double time_taken = chrono::duration_cast<chrono::nanoseconds>(end - start).count();
    time_taken *= 1e-9;
    cout << "Time taken by program is : " << endl
         << fixed
         << time_taken << setprecision(9) << " seconds" << endl;
#endif
    return 0;
}